base/invalidation.jl

```diff
@@ -48,83 +45,138 @@ function foreach_module_mtable(visit, m::Module, world::UInt)
+    return false
+end
+
+function should_invalidate_code_for_globalref(gr::GlobalRef, src::CodeInfo)
+    isgr(g::GlobalRef) = gr.mod == g.mod && gr.name === g.name
+    isgr(g) = false
+    return anygr(isgr, src)
 end
 
-function scan_edge_list(ci::Core.CodeInstance, bpart::Core.BindingPartition)
+function scan_edge_list(ci::Core.CodeInstance, binding::Core.Binding)
     isdefined(ci, :edges) || return false
     edges = ci.edges
     i = 1
     while i <= length(edges)
-        if isassigned(edges, i) && edges[i] === bpart
+        if isassigned(edges, i) && edges[i] === binding
             return true
         end
         i += 1
     end
     return false
 end
 
+function invalidate_method_for_globalref!(gr::GlobalRef, method::Method, invalidated_bpart::Core.BindingPartition, new_max_world::UInt)
+    if isdefined(method, :source)
+        src = _uncompressed_ir(method)
+        binding = convert(Core.Binding, gr)
+        old_stmts = src.code
+        invalidate_all = should_invalidate_code_for_globalref(gr, src)
+        for mi in specializations(method)
+            isdefined(mi, :cache) || continue
+            ci = mi.cache
+            while true
+                if ci.max_world > new_max_world && (invalidate_all || scan_edge_list(ci, binding))
+                    ccall(:jl_invalidate_code_instance, Cvoid, (Any, UInt), ci, new_max_world)
+                end
+                isdefined(ci, :next) || break
+                ci = ci.next
+            end
+        end
+    end
+end
+
 function invalidate_code_for_globalref!(gr::GlobalRef, invalidated_bpart::Core.BindingPartition, new_max_world::UInt)
     try
         valid_in_valuepos = false
-        foreach_reachable_mtable(new_max_world) do mt::Core.MethodTable
+        foreach_module_mtable(gr.mod, new_max_world) do mt::Core.MethodTable
             for method in MethodList(mt)
-                if isdefined(method, :source)
-                    src = _uncompressed_ir(method)
-                    old_stmts = src.code
-                    invalidate_all = should_invalidate_code_for_globalref(gr, src)
-                    for mi in specializations(method)
-                        isdefined(mi, :cache) || continue
-                        ci = mi.cache
-                        while true
-                            if ci.max_world > new_max_world && (invalidate_all || scan_edge_list(ci, invalidated_bpart))
-                                ccall(:jl_invalidate_code_instance, Cvoid, (Any, UInt), ci, new_max_world)
-                            end
-                            isdefined(ci, :next) || break
-                            ci = ci.next
-                        end
-                    end
-                end
+                invalidate_method_for_globalref!(gr, method, invalidated_bpart, new_max_world)
             end
             return true
         end
+        b = convert(Core.Binding, gr)
+        if isdefined(b, :backedges)
+            for edge in b.backedges
```

Comment:
This appears to trigger UB when julia runs with threads (the default) and does any inference, since `b.backedges` is accessed without a lock here